### PR TITLE
Fluid/feature dynamic subscales

### PR DESCRIPTION
**Pull request by jcotela**
porting the dynamic subscale element to the new fluid_element framework.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_elements/d_vms.h [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    constexpr static double mSubscalePredictionVelocityTolerance = 1e-14;
+    constexpr static double mSubscalePredictionResidualTolerance = 1e-14;
+    constexpr static unsigned int mSubscalePredictionMaxIterations = 10;
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // Velocity subscale history, stored at integration points
+    std::vector< array_1d<double,Dim> > mPredictedSubscaleVelocity;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/d_vms.h:326]: is it deliberate to employ std::vector here?

File: applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.cpp [hunk 2]
```diff
@@ -172,6 +172,55 @@ void FluidElementUtilities<TNumNodes>::SetTangentialProjectionMatrix(
...
 }
 
+
+template < std::size_t TNumNodes>
+void FluidElementUtilities<TNumNodes>::DenseSystemSolve(
+    const boost::numeric::ublas::bounded_matrix<double,2,2> &rA,
+    const array_1d<double,2> &rB,
+    array_1d<double,2> &rX)
+{
+    boost::numeric::ublas::bounded_matrix<double,2,2> inverse;
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.cpp:182]: i would suggest to remove the boost::numeric::etc...

but it may well be a different PR

Comment (RiccardoRossi) [reply applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.cpp:182]: why don't you use the one in math_utils btw?

File: applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.h [hunk 3]
```diff
@@ -189,6 +189,28 @@ class FluidElementUtilities {
         const array_1d<double, 3>& rUnitNormal,
         boost::numeric::ublas::bounded_matrix<double, 3, 3>& rTangProjMatrix);
 
+    /**
+     * Invert a system with 2 unknowns, defined using bounded matrix types.
+     * @param[in] rA the 2x2 system matrix.
+     * @param[in] rB the right hand side vector.
+     * @param[out] rX the solution of the system.
+     */
+    static void DenseSystemSolve(
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.h:198]: math_utils

File: applications/FluidDynamicsApplication/tests/fluid_element_test.py [hunk 4]
```diff
@@ -182,30 +212,36 @@ def checkResults(self):
...
+        self.gid_io.FinalizeResults()
 
 if __name__ == '__main__':
     test = FluidElementTest()
     test.setUp()
     test.print_reference_values = False
     test.print_output = True
-    test.testCavityOSS()
-    #test.testCavity()
+    #test.testCavityQSASGS()
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/tests/fluid_element_test.py:243]: why are u taking out the existing one? is it intentional?

**Review by RiccardoRossi — APPROVED** [review]
minor comments. @rubenzorrilla could u also take a look? i am approving as i don't see it harmful, however i wonder if some of the changes in the FluidElement might affect you (easy fix though)

File: applications/FluidDynamicsApplication/tests/fluid_element_test.py (hunk 4 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/tests/fluid_element_test.py:243]: yes, the existing one has been renamed (it should be testCavityQSASGS now)

File: applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.cpp (hunk 2 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.cpp:182]: @RiccardoRossi I'm merging now to do this renaming all in one go ;)
Regarding the math utils thing, this is just adapted from my old element, which used its own function for some reason (I think back then there was no bounded matrix version of this function in math utils, but I might be wrong). I can refactor it as part of the porting to AMatrix.

File: applications/FluidDynamicsApplication/custom_elements/d_vms.h (hunk 1 above)
Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_elements/d_vms.h:326]: yes, it is variable size (num gauss points) and it is not expected to be used it in combination with matrix operations. Which container would you suggest?

**Comment by jcotela** [comment]
@RiccardoRossi  @rubenzorrilla any changes to the existing elements have been propagated, and the tests pass. I am more concerned by the ongoing work of @ddiezrod and @ipouplana, since those are not in the repository yet. Feel free to contact me if you have any kind of trouble, though.

File: applications/FluidDynamicsApplication/custom_elements/d_vms.h (hunk 1 above)
Comment (RiccardoRossi) [reply applications/FluidDynamicsApplication/custom_elements/d_vms.h:326]: i would have suggested the ublas::vector. You would need to do a resize instead of push back, but i think it is slightly faster. also it occupies less memory since it does not allocate "capacity" items

Comment (jcotela) [reply applications/FluidDynamicsApplication/custom_elements/d_vms.h:326]: should I go for the amatrix dense vector then? I can change it as I refactor.
